### DISCUSSION TITLE
Install applications in workers

### DISCUSSION THREAD

**Discussion by karimDev13** [alibaba/GraphScope#1980] [General]
Please i've a question , can i install other applications in each worker on graphscope ?

**Comment by sighingnow** [comment]
Definitely yes.

You could build a docker image from the graphscope's image: `registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:0.16.0`.

e.g.,

```Dockerfile

FROM registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:0.16.0

RUN pip3 install ...
```

And then use your own image to launch graphscope session by:

```python
session = graphscope.session("k8s_gs_image=...")
```

See also: https://github.com/alibaba/GraphScope/blob/main/python/graphscope/client/session.py#L297-L314

**Reply by karimDev13** [reply]
@sighingnow thank you very much for your answer, but concerning the connections between the new application and graphscope how is it done (especially in the case of parallelism of many workers) ?

**Reply by sighingnow** [reply]
Hi @karimDev13,

May I know more details about your "applications"? e.g., which component your application want to connect to? Vineyard?

**Reply by karimDev13** [reply]
@sighingnow  the application is a new interactive engine (with python library) that must be connected to each worker.

**Reply by sighingnow** [reply]
> that must be connected to each worker.

Which worker? GAE or GIE or GLE?

**Reply by karimDev13** [reply]
graph interactive engine (GIE) exactly each node containing fragment of the partitioned graph instance must be connected to this interactive engine for parallel execution.

**Reply by sighingnow** [reply]
How would you like to connect to the GIE worker process inside the pod? As far as I know GIE engine process doesn't expose something like TCP socket or HTTP server to allow establishing external connections.
